### PR TITLE
bridge: Ignore network traffic on loopback

### LINKED ISSUE #7338: network load graph includes loopback traffic
As part of the investigations in issue #7318 I noticed that our "Network Traffic" graph on /system includes traffic on "lo" (i. e. the loopback device on localhost). This does not appear to be very useful, it would be better to exclude it.

For that we would need to find a better way of generating traffic in test/verify/check-metrics; possibly with generating a veth pair and piping our /dev/zero stream through that.

### PR DESCRIPTION
**Pull request by martinpitt**
For most use cases loopback traffic is not very interesting, and even
misleading for server administration. It is also inconsistent with the 
graphs on "Networking" and "Dashboard", as these also ignore the lo
interface.

Move from the network.all.{tx,rx} internal metric towards the 
instantiated network.interface.{tx,rx} internal metric, as that can be
influenced with "omit-instances" and mirrors what the PCP metrics do
(network.interface.total.bytes also has per-interface instances). The 
per-interface measurements are done anyway, and the summation of the 
instances is not much extra overhead (and done for PCP too).

Drop the now unused network.all.{tx,rx} internal metrics.

Adjust check-metrics to move the netcat traffic to the Ethernet
interface by putting the sending side outside of the test VM. This
requires forwarding port 2000 and opening it in the firewall (where
applicable). Add another check that generates 5s of traffic on localhost
(via flood ping) and ensure that the networking graph does *not* exceed
300 kbps.

Fixes #7338

### CODE REVIEW

**Comment by martinpitt** [comment]
This works and stops the /system, /network, and /dashboard pages having three (!) different opinions about loopback traffic counting - it is now consistently ignored everywhere.

That said, I'm not entirely happy with this fix, hardcoding the `lo` ignoring feels a bit odd. IMHO it would be better to teach the internal metrics to actually pay attention to `omit-instances` (it currently ignores that), drop the `network.all.{rx,tx}` internal metrics entirely, and use the `network.interface.{rx,tx}` instance groups instead (with filtering out `lo`). This will make the internal network metrics work the same way as the PCP ones. I look into this.

**Comment by martinpitt** [comment]
I reworked this as above; `omit-instances` now works with internal metrics, I added an unit test case for this. This spotted a memory leak, fixed in a separate commit. With that we can get rid of `network.all.{tx,rx}` and use the per-interface instantiated metrics and filter out `lo` cleanly.

File: src/bridge/cockpitinternalmetrics.c [hunk 1]
```diff
@@ -242,6 +240,11 @@ cockpit_internal_metrics_sample (CockpitSamples *samples,
 {
   CockpitInternalMetrics *self = COCKPIT_INTERNAL_METRICS (samples);
 
+  if (self->omit_instances)
```
Comment (martinpitt) [inline src/bridge/cockpitinternalmetrics.c]: OK, added locally. Waiting a bit with the push in case you have some other feedback, to not unnecessarily run CI.

Comment (martinpitt) [inline src/bridge/cockpitinternalmetrics.c]: Pushed.

Comment (petervo) [inline src/bridge/cockpitinternalmetrics.c]: I think the for loop needs braces too. Generally any block that is more than one line should have them.

File: src/bridge/cockpitnetworksamples.c [hunk 2]
```diff
@@ -92,14 +89,8 @@ cockpit_network_samples (CockpitSamples *samples)
 
       cockpit_samples_sample (samples, "network.interface.rx", iface_name, bytes_rx);
       cockpit_samples_sample (samples, "network.interface.tx", iface_name, bytes_tx);
-
-      total_rx += bytes_rx;
-      total_tx += bytes_tx;
     }
 
-  cockpit_samples_sample (samples, "network.all.rx", NULL, total_rx);
-  cockpit_samples_sample (samples, "network.all.tx", NULL, total_tx);
```
Comment (stefwalter) [inline src/bridge/cockpitnetworksamples.c]: Isn't this (borderline) breaking API? Are we sure we're OK with that? @petervo what do you think?

File: src/bridge/cockpitinternalmetrics.c (hunk 1 above)
Comment (martinpitt) [inline src/bridge/cockpitinternalmetrics.c]: Hmm, seems a bit excessive to me, but updated.

File: src/bridge/cockpitnetworksamples.c (hunk 2 above)
Comment (martinpitt) [inline src/bridge/cockpitnetworksamples.c]: We don't document the available metrics, and a google search for these two didn't turn up anything except our own code. Formally it's API, sure, but it's not even a useful metric really (and differs to what you get from PCP). Under these considerations I felt it is okay to remove it, but if you prefer to keep it, we could do that and just not use it ourselves any more.

File: src/bridge/cockpitinternalmetrics.c [hunk 3]
```diff
@@ -81,8 +81,6 @@ static MetricDescription metric_descriptions[] = {
   { "disk.all.read",    "bytes", "counter", FALSE, DISK_SAMPLER },
   { "disk.all.written", "bytes", "counter", FALSE, DISK_SAMPLER },
 
-  { "network.all.rx",       "bytes", "counter", FALSE, NETWORK_SAMPLER },
```
Comment (petervo) [inline src/bridge/cockpitinternalmetrics.c]: I think we need to leave these, so that old shells can work the same as they work today with the new bridge.

Comment (stefwalter) [inline src/bridge/cockpitinternalmetrics.c]: For better or worse, I agree. This is what we've signed up for with the compatibility between our Cockpit components. In addition if we're not using this functionality, we should write a basic unit test for it, to make sure these options are still recognized.

Comment (martinpitt) [inline src/bridge/cockpitinternalmetrics.c]: Ah sorry, I didn't think about that case. Reworked to keep these; I added a "deprecated" comment and a unit test.

**Comment by martinpitt** [comment]
@petervo: I don't think we need to bump it. `network.interface.*` was  added in https://github.com/cockpit-project/cockpit/commit/cc547709e6abf in version 0.89, and systemd already depends on bridge 122, and dashboard on 138. If you refer to `omit-instances` support, this isn't strictly necessary -- older bridges just ignore it, so you'll get loopback traffic included, but that happens with the current version also: At the time of above commit (0.89) `network.all.{tx,rx}` still filtered out loopback traffic, but that was wrongly dropped in https://github.com/cockpit-project/cockpit/commit/7541a88bf18613d in version 0.96, so all cockpit-bridge versions that are out there (in back branches) include loopback traffic.

So in summary, I'd rather leave the dependencies as they are. Sure you'll get better behaviour with newer versions, but it's not any worse with old ones.

**Comment by stefwalter** [comment]
@martinpitt  The logic makes sense to me.
